README.md
# SN II explosion time from optical spectra

The explosion time t_0 of an SN can be estimated by means of photometric information; it can be constrained between the last non-detection and the first detection, or estimated through a polynomial fit to the rise-time photometry when it is available. The spectroscopy of an SN can also provide information about its explosion epoch by means of the comparison with other spectra of SNe with explosion epoch estimated through photometric information.

In [Rodríguez et al. (2019)](https://ui.adsabs.harvard.edu/abs/2019MNRAS.483.5459R/abstract) I developed an automated routine to compute the probability density function (pdf) of t_0 given a set of input optical spectra, called **SNII Explosion Time from Optical Spectra (SNII_ETOS)**. This routine makes use of the Supernova Identificator code (SNID; [Blondin & Tonry 2007](https://ui.adsabs.harvard.edu/abs/2007ApJ...666.1024B/abstract)), which finds by cross-correlation the spectra from a certain template library that are more similar to the input spectrum. In [Rodríguez et al. (2019)](https://ui.adsabs.harvard.edu/abs/2019MNRAS.483.5459R/abstract) I also compiled a template library using optical spectra of 59 SNe II with t_0 constrained within 10 days by photometric information (details of the template library are available [here](https://github.com/olrodrig/SNII_ETOS/tree/master/docs/R19_tableF6.pdf)). These spectra are at epochs earlier than 40 days since explosion and have information in the range between 4100 and 7000 angstrom. Therefore, to compute t_0 with the **SNII_ETOS** code it is necessary that the input spectra: (1) be defined between 4100 and 7000 angstrom, and (2) be as early as possible (at least 40 days earlier than the first detection).

For any question, email me at olrodrig@gmail.com

**If you use the SNII_ETOS code in your work, please cite [Rodríguez et al. (2019)](https://ui.adsabs.harvard.edu/abs/2019MNRAS.483.5459R/abstract).**


CHANGELOG
2020/04/25: version 1.0


src/plot_ETOS.py
import os
import matplotlib
import matplotlib.pyplot as plt
from pdf_routines import gaussian_pdf
import matplotlib.ticker as plticker

matplotlib.rc('axes', labelsize=15.0)
textsize = 11

def draw_ticks(ax, i, x_major, x_minor):
    
  ax.yaxis.set_major_formatter(plticker.NullFormatter())
  ax.yaxis.set_ticks_position('none') 
  
  ax.xaxis.set_major_locator(plticker.MultipleLocator(base=x_major))
  ax.xaxis.set_minor_locator(plticker.MultipleLocator(base=x_minor))
  
  ax.tick_params('x', length=6, width=1, which='major', direction='in')
  ax.tick_params('x', length=3, width=1, which='minor', direction='in')
  ax.xaxis.set_ticks_position('both')
  
  ax.set_ylabel('pdf($t_0$)', labelpad=-1)
  
def color_line_style(templates):
  colors = ['b','lime','r','c','m','y','gold','maroon','darkgreen','darkgray','violet','darkkhaki','teal']
  lines  = ['-','--',':','-.']
  template_color, template_line, i_color, i_line = {}, {}, 0, 0
  for i in range(0, len(templates)):
      template_color[templates[i]] = colors[i_color]
      template_line[templates[i]]  = lines[i_line]
      i_color = i_color+1
      if i_color == len(colors):
          i_color = 0
          i_line  = i_line+1
  return template_color, template_line

def plot_gaussian_pdfs(sne_bm, t0s_bm, rms_t0s_bm, x_pdf, ax, xpos, ypc, text, template_line, template_color, textsize, panel, with_labels=False, label_pos=0.0):
  ymax = []
  for sn_bm, spec_phase, err_spec_phase in zip(sne_bm, t0s_bm, rms_t0s_bm):
      g_per_spec = gaussian_pdf(spec_phase, err_spec_phase, x_pdf)
      ax.plot(x_pdf, g_per_spec, linestyle=template_line[sn_bm], color=template_color[sn_bm], label=sn_bm)
      ymax.append(max(g_per_spec))
  ymax = max(ymax)
  ax.set_ylim(0, ymax*1.1)
  ax.text(xpos, ymax*ypc, text, fontsize=textsize)
  ax.text(-54, ymax*ypc, panel, fontsize=textsize)
  xmin, xmax = min(x_pdf), max(x_pdf)
  ax.set_xlim(xmin, xmax)
  if with_labels:  ax.legend(bbox_to_anchor=(0.0, label_pos, 1.0, 1.0), loc=3, ncol=8, borderaxespad=0.0, labelspacing=0.4, handletextpad=0.2, mode='expand', numpoints=1)
  
def plot_non_gaussian_pdfs(sne_bm, pdfs_per_sn, x_pdf, ax, xpos, ypc, text, template_line, template_color, textsize, panel):
  ymax = []
  for sn_bm, pdf_per_sn in zip(sne_bm, pdfs_per_sn):
      ax.plot(x_pdf, pdf_per_sn, linestyle=template_line[sn_bm], color=template_color[sn_bm], label=sn_bm)
      ymax.append(max(pdf_per_sn))
  ymax = max(ymax)
  ax.set_ylim(0, ymax*1.1)
  ax.text(xpos, ymax*ypc, text, fontsize=textsize)
  ax.text(-54, ymax*ypc, panel, fontsize=textsize)
  xmin, xmax = min(x_pdf), max(x_pdf)
  ax.set_xlim(xmin, xmax)

def plot_final_pdfs(pdf, x_pdf, JD_ln, n_spec, ax):
  if JD_ln != -9999:  ax.plot(x_pdf, pdf['fd+ln'], '-', color='dimgray', lw=2)
  ax.plot(x_pdf, pdf['snid'], '-k', lw=3)
  ax.plot(x_pdf, pdf['snid+fd'], '-', color='lime', lw=1)
  if JD_ln != -9999:  ax.plot(x_pdf, pdf['snid+fd+ln'], '--', color='r', lw=1)
  
  ymax = max([max(pdf['fd+ln']),max(pdf['snid']),max(pdf['snid+fd']),max(pdf['snid+fd+ln'])])
  ax.set_ylim(0, ymax*1.1)
  xpos = 14
  if JD_ln != -9999:  ax.text(xpos, ymax*0.82, "Uniform pdf between JD$_\mathrm{fd}$ and JD$_\mathrm{ln}$", color='dimgray', fontsize=textsize)
  if n_spec == 1:
      ax.text(xpos, ymax*0.65, "SNID pdf (+ a rms error of 5.0 days)", color='k', fontsize=textsize)
  else:
      ax.text(xpos, ymax*0.65, "SNID pdf (+ a rms error of 4.1 days)", color='k', fontsize=textsize)
  ax.text(xpos, ymax*0.48, "SNID pdf with JD$_\mathrm{fd}$ as prior", color='lime', fontsize=textsize)
  if JD_ln != -9999:  ax.text(xpos, ymax*0.31, "SNID pdf with JD$_\mathrm{fd}$ and JD$_\mathrm{ln}$ as priors", color='r', fontsize=textsize)
  ax.set_xlim(min(x_pdf), max(x_pdf))
  ax.set_xlabel('$(t-\mathrm{JD}_\mathrm{fd}$)/($1+z$) [day]')
  
  ax.text(-54, ymax*0.82, '(e)', fontsize=textsize)

def plot_ETOS(ETOS, figure_name):
  
  sn     = ETOS.sn
  n_spec = ETOS.n_spec
  sne_bm          = ETOS.for_plots[0]
  t0s_bm          = ETOS.for_plots[1]
  rms_t0s_bm      = ETOS.for_plots[2]
  pdfs_snid_sn_bm = ETOS.for_plots[3]
  x_pdf = ETOS.x_pdf
  pdf   = ETOS.pdf
  
  #define colors and line styles
  template_color, template_line = color_line_style(sne_bm['3'])
  
  fig = plt.figure(figsize=(8, 6))
  if len(sne_bm['3']) <= 8:
      fig.subplots_adjust(wspace=0.0, hspace=0.0, left=0.04, bottom=0.075, right=0.99, top=0.92)
  elif len(sne_bm['3']) <= 16:
      fig.subplots_adjust(wspace=0.0, hspace=0.0, left=0.04, bottom=0.075, right=0.99, top=0.889)
  else:
      fig.subplots_adjust(wspace=0.0, hspace=0.0, left=0.04, bottom=0.075, right=0.99, top=0.858)
      
  ax = {}
  for i in range(1, 6):  ax[str(i)] = fig.add_subplot(5,1,i)

  plot_gaussian_pdfs(sne_bm['1'], t0s_bm['1'], rms_t0s_bm['1'], x_pdf, ax['1'], 19, 0.82, "all pdf's per SN$_\mathrm{bm}$ per spectrum", template_line, template_color, textsize, '(a)')
  plot_gaussian_pdfs(sne_bm['2'], t0s_bm['2'], rms_t0s_bm['2'], x_pdf, ax['2'], 16, 0.82, "typical pdf per SN$_\mathrm{bm}$ per spectrum", template_line, template_color, textsize, '(b)')
  plot_gaussian_pdfs(sne_bm['3'], t0s_bm['3'], rms_t0s_bm['3'], x_pdf, ax['3'], 35, 0.82, "<pdf> per SN$_\mathrm{bm}$",template_line,template_color, textsize, '(c)', with_labels=True,label_pos=3.02)
  plot_non_gaussian_pdfs(sne_bm['3'], pdfs_snid_sn_bm, x_pdf, ax['4'], 20, 0.82, "<pdf> per SN$_\mathrm{bm}$, with $t_0$ error", template_line, template_color, textsize, '(d)')
  plot_final_pdfs(pdf, x_pdf, ETOS.JD_ln, n_spec, ax['5'])
   
  for i in range(1, 6):  draw_ticks(ax[str(i)], i, 10.0, 2.0)
  
  if n_spec == 1:
      title = 'SN '+sn+'; 1 input spectrum'
  else:
      title = 'SN '+sn+'; '+str(n_spec)+' input spectra'
  if len(sne_bm['3']) <= 8:
      ax['1'].set_title(title, position=(0.5,1.25))
  elif len(sne_bm['3']) <= 16:
      ax['1'].set_title(title, position=(0.5,1.45))
  else:
      ax['1'].set_title(title, position=(0.5,1.65))
  
  if figure_name != '':  fig.savefig(figure_name+'.pdf', dpi=300)


src/pdf_routines.py
import numpy as np
from sklearn import mixture

def snii_templates_epochs():
    
  JD_ln_template, JD_fd_template = {}, {}
  JD_ln_template['1986L']  , JD_fd_template['1986L']   = 2446705.5  ,  2446711.1
  JD_ln_template['1990E']  , JD_fd_template['1990E']   = 2447932.5  ,  2447937.62
  JD_ln_template['1999br'] , JD_fd_template['1999br']  = 2451272.9  ,  2451280.9
  JD_ln_template['1999em'] , JD_fd_template['1999em']  = 2451471.95 ,  2451479.51
  JD_ln_template['1999gi'] , JD_fd_template['1999gi']  = 2451515.68 ,  2451522.32
  JD_ln_template['1999go'] , JD_fd_template['1999go']  = 2451527.7  ,  2451535.7
  JD_ln_template['2000dc'] , JD_fd_template['2000dc']  = 2451758.8  ,  2451765.8
  JD_ln_template['2000dj'] , JD_fd_template['2000dj']  = 2451785.487,  2451795.9
  JD_ln_template['2000el'] , JD_fd_template['2000el']  = 2451835.7  ,  2451840.6
  JD_ln_template['2001X']  , JD_fd_template['2001X']   = 2451958.0  ,  2451968.3
  JD_ln_template['2001do'] , JD_fd_template['2001do']  = 2452131.7  ,  2452135.7
  JD_ln_template['2001fa'] , JD_fd_template['2001fa']  = 2452195.9  ,  2452200.9
  JD_ln_template['2002an'] , JD_fd_template['2002an']  = 2452292.04 ,  2452297.02
  JD_ln_template['2002ce'] , JD_fd_template['2002ce']  = 2452369.7  ,  2452375.378
  JD_ln_template['2002gd'] , JD_fd_template['2002gd']  = 2452549.28 ,  2452550.53
  JD_ln_template['2003Z']  , JD_fd_template['2003Z']   = 2452660.2  ,  2452669.2
  JD_ln_template['2003bn'] , JD_fd_template['2003bn']  = 2452691.5  ,  2452692.83
  JD_ln_template['2003ej'] , JD_fd_template['2003ej']  = 2452770.8  ,  2452779.8
  JD_ln_template['2003hg'] , JD_fd_template['2003hg']  = 2452860.9  ,  2452869.9
  JD_ln_template['2003hl'] , JD_fd_template['2003hl']  = 2452863.0  ,  2452872.0
  JD_ln_template['2003iq'] , JD_fd_template['2003iq']  = 2452918.47 ,  2452921.458
  JD_ln_template['2004ci'] , JD_fd_template['2004ci']  = 2453168.9  ,  2453171.8
  JD_ln_template['2004er'] , JD_fd_template['2004er']  = 2453269.88 ,  2453273.9
  JD_ln_template['2004et'] , JD_fd_template['2004et']  = 2453270.517,  2453271.483
  JD_ln_template['2004fc'] , JD_fd_template['2004fc']  = 2453292.89 ,  2453295.124
  JD_ln_template['2004fx'] , JD_fd_template['2004fx']  = 2453300.92 ,  2453306.93
  JD_ln_template['2005ay'] , JD_fd_template['2005ay']  = 2453449.121,  2453456.58
  JD_ln_template['2005cs'] , JD_fd_template['2005cs']  = 2453548.43 ,  2453549.41
  JD_ln_template['2005dz'] , JD_fd_template['2005dz']  = 2453615.8  ,  2453623.71
  JD_ln_template['2006Y']  , JD_fd_template['2006Y']   = 2453763.09 ,  2453770.08
  JD_ln_template['2006bc'] , JD_fd_template['2006bc']  = 2453811.087,  2453819.15
  JD_ln_template['2006bp'] , JD_fd_template['2006bp']  = 2453833.677,  2453834.647
  JD_ln_template['2006it'] , JD_fd_template['2006it']  = 2454004.69 ,  2454009.67
  JD_ln_template['2006iw'] , JD_fd_template['2006iw']  = 2454009.737,  2454011.798
  JD_ln_template['2007hv'] , JD_fd_template['2007hv']  = 2454342.5  ,  2454352.87
  JD_ln_template['2007il'] , JD_fd_template['2007il']  = 2454345.94 ,  2454353.95
  JD_ln_template['2007pk'] , JD_fd_template['2007pk']  = 2454409.83 ,  2454414.81
  JD_ln_template['2008bh'] , JD_fd_template['2008bh']  = 2454538.57 ,  2454548.66
  JD_ln_template['2008br'] , JD_fd_template['2008br']  = 2454559.323,  2454564.265
  JD_ln_template['2008ho'] , JD_fd_template['2008ho']  = 2454787.77 ,  2454796.61
  JD_ln_template['2008if'] , JD_fd_template['2008if']  = 2454802.73 ,  2454812.71
  JD_ln_template['2008il'] , JD_fd_template['2008il']  = 2454822.69 ,  2454827.64
  JD_ln_template['2008in'] , JD_fd_template['2008in']  = 2454824.45 ,  2454824.95
  JD_ln_template['2009ao'] , JD_fd_template['2009ao']  = 2454886.62 ,  2454894.62
  JD_ln_template['2009bz'] , JD_fd_template['2009bz']  = 2454912.03 ,  2454919.98
  JD_ln_template['2010id'] , JD_fd_template['2010id']  = 2455450.82 ,  2455454.743
  JD_ln_template['2012aw'] , JD_fd_template['2012aw']  = 2456001.769,  2456003.349
  JD_ln_template['2013am'] , JD_fd_template['2013am']  = 2456371.698,  2456373.138
  JD_ln_template['2013by'] , JD_fd_template['2013by']  = 2456402.872,  2456403.752
  JD_ln_template['2013ej'] , JD_fd_template['2013ej']  = 2456497.04 ,  2456497.625
  JD_ln_template['2013fs'] , JD_fd_template['2013fs']  = 2456570.82 ,  2456571.737
  JD_ln_template['2013hj'] , JD_fd_template['2013hj']  = 2456635.7  ,  2456638.8
  JD_ln_template['2014G']  , JD_fd_template['2014G']   = 2456668.35 ,  2456671.111
  JD_ln_template['LSQ14gv'], JD_fd_template['LSQ14gv'] = 2456670.7  ,  2456674.8
  JD_ln_template['2014cx'] , JD_fd_template['2014cx']  = 2456901.89 ,  2456902.90
  JD_ln_template['2014cy'] , JD_fd_template['2014cy']  = 2456898.8  ,  2456900.5
  JD_ln_template['2015bs'] , JD_fd_template['2015bs']  = 2456915.5  ,  2456925.5
  JD_ln_template['2016esw'], JD_fd_template['2016esw'] = 2457607.802,  2457608.814

  return JD_ln_template, JD_fd_template

#compute weighted average
def weighted_average(x, sigma_x, with_intrinsic_error=True):
    
  if len(x) > 1:
      if with_intrinsic_error:
          residuals = x - np.mean(x)
          rms       = np.sqrt(np.sum(residuals**2)/float(len(residuals)-1))
          sigma_0s  = np.linspace(0.0, rms, 100)
      else:
          sigma_0s  = np.array([0.0])
          
      m2lnL_min = 1.e90
      for sigma_0 in sigma_0s:
          
          Var   = sigma_x**2 + sigma_0**2
          w_ave = np.sum(x/Var)/np.sum(1.0/Var)
          m2lnL = np.sum(np.log(Var)+(x-w_ave)**2/Var)
          
          if m2lnL < m2lnL_min:
              m2lnL_min = m2lnL
              best_x    = w_ave
              best_error = np.sqrt(1.0/np.sum(1.0/Var))
  else:
      best_x, best_error = x[0], sigma_x[0]
      
  return best_x, best_error

#pick rangom values given a pdf
def values_from_distribution(x, pdf, N):

  x_sample = np.random.choice(x, N, p=pdf/np.sum(pdf)) #sum of probabilities must to be 1

  return x_sample

#Simpson's rule
def simpson(x,f):
  
  integral = (f[0] + f[-1]) / 3.0  #extremes
  n = len(x)
  
  four = "o"
  for i in range(1, n - 1):
      if four == "o":
          integral += f[i] * 4.0 / 3.0
          four = "x"
      else:
          integral += f[i] * 2.0 / 3.0
          four = "o"
              
  integral = (x[1] - x[0]) * integral
      
  return integral

#discard possible outliers through the Tukey's rule
def tukey_rule(x, k=1.5):

  Q1, Q3 = np.quantile(x, [0.25, 0.75])
  IQR    = Q3 - Q1
  
  x = x[x>=Q1-k*IQR]
  x = x[x<=Q3+k*IQR]
  
  return x

#return a normalized gaussian pdf
def gaussian_pdf(mu, sigma, x_sampled):
  g_sampled = np.exp(-0.5*(mu-x_sampled)**2/sigma**2)
  g_sampled = g_sampled / simpson(x_sampled, g_sampled)
  return g_sampled

#return a uniform pdf
def uniform_pdf(x_min, x_max, x):
  h = 1.0/(x_max-x_min)
  pdf = np.linspace(h, h, len(x))
  for i in range(0, len(x)):
      if x[i] < x_min or x[i] > x_max:
          pdf[i] = 0.0
  return pdf

#return a pdf computed as a mixture of Gaussians
def get_pdf(y, y_sampled, max_components=2):
  
  x, x_sampled = y.reshape(-1,1), y_sampled.reshape(-1,1)
  
  BIC_min = 1.e90
  for n_components in range(1, max_components+1):
      gmm   = mixture.GaussianMixture(n_components=n_components)
      model = gmm.fit(x)
      BIC   = model.bic(x)
      if BIC < BIC_min:
          BIC_min   = BIC
          model_min = model         
  ln_pdf = model_min.score_samples(x_sampled)
  pdf    = np.exp(ln_pdf)
  
  return pdf

#return different pdf's
def final_pdfs(z, JD_ln, JD_fd, pdfs_per_sn, x_sampled, N_sample, rms_t0):
    
  #define the uniform pdf's given by the JD_fd and JD_fd+JD_ln
  ln, fd    = (JD_ln - JD_fd)/(1.0+z), 0.0
  pdf_fd    = uniform_pdf(-9999.0, fd, x_sampled) #fd as prior
  pdf_fd_ln = uniform_pdf(ln, fd, x_sampled) #fd and ln as prior
    
  #combine the pdf of different sne
  pdf_snid = np.linspace(1.0, 1.0, len(x_sampled))  
  for pdf_per_sn in pdfs_per_sn:
      pdf_snid = pdf_snid*pdf_per_sn
  
  #add typical rms(t0) error
  err_0    = np.random.normal(0.0, rms_t0, N_sample)
  err_0    = np.random.choice(tukey_rule(err_0), N_sample)
  err_0    = err_0 - np.median(err_0)
  t0s_snid = values_from_distribution(x_sampled, pdf_snid, N_sample)
  t0s_snid = t0s_snid + err_0
  t0s_snid = np.random.choice(tukey_rule(t0s_snid),N_sample)
  
  #compute pdf's
  pdf_snid       = get_pdf(t0s_snid, x_sampled, max_components=1)
  pdf_snid_fd    = pdf_snid*pdf_fd
  pdf_snid_fd_ln = pdf_snid*pdf_fd_ln
  
  #normalize pdf's
  pdf_snid       = pdf_snid / simpson(x_sampled, pdf_snid)
  pdf_snid_fd    = pdf_snid_fd / simpson(x_sampled, pdf_snid_fd)
  pdf_snid_fd_ln = pdf_snid_fd_ln / simpson(x_sampled, pdf_snid_fd_ln)
  
  return pdf_fd_ln, pdf_snid, pdf_snid_fd, pdf_snid_fd_ln


def average_pdf_per_sn_bm_with_t0_error(sne_bm, t0s_bm, rms_t0s_bm, x_pdf, N_sample):
  
  JD_ln_template, JD_fd_template = snii_templates_epochs()
  
  pdfs_per_sn = []
  for sn_bm, spec_phase, err_spec_phase in zip(sne_bm, t0s_bm, rms_t0s_bm):
            
      delta = round(JD_fd_template[sn_bm]-JD_ln_template[sn_bm],3)
      rms_uniform = delta/np.sqrt(12.0)
      
      if rms_uniform < 0.3*err_spec_phase:
          pdf_per_sn = gaussian_pdf(spec_phase, err_spec_phase, x_pdf)
      else:
          err_t0_template = np.random.uniform(-0.5*delta,0.5*delta, N_sample)
          err_t0_template = err_t0_template - np.median(err_t0_template) #center the distibution to zero
          
          x1 = np.random.normal(spec_phase, err_spec_phase, N_sample)
          x1 = np.random.choice(tukey_rule(x1), N_sample)
          x1 = x1 - np.median(x1) + spec_phase #center the distibution to the phase
          #include values from the uniform distrution
          x= x1 + err_t0_template
          pdf_per_sn = get_pdf(x, x_pdf)
          pdf_per_sn = pdf_per_sn / simpson(x_pdf, pdf_per_sn)
      pdfs_per_sn.append(pdf_per_sn)

  return pdfs_per_sn

def average_pdf_per_sn_bm(t0s_best, rms_t0s_best, sne_best):
    
  #best matching SNe
  sne_bm = list(set(sne_best))

  t0s_bm, rms_t0s_bm = [], []
  for sn_bm in sne_bm:
      phases, err_phases = np.array([]), np.array([])
      for sn_i, spec_phase, err_spec_phase in zip(sne_best, t0s_best, rms_t0s_best):
          if sn_i == sn_bm:
              phases     = np.append(phases, spec_phase)
              err_phases = np.append(err_phases, err_spec_phase)
      t0_best, rms_t0_best =  weighted_average(phases, err_phases)
      t0s_bm.append(t0_best)
      rms_t0s_bm.append(rms_t0_best)
    
  return sne_bm, t0s_bm, rms_t0s_bm

def typical_pdf_per_sn_bm_per_spectrum(t0s_best, rms_t0s_best, sne_best, t_spec_best):

  epochs = list(set(t_spec_best))
  
  new_sne_best, new_t0_best, new_rms_t0_best = [], [], []
  for epoch in epochs:
      
      #number of templates at the epoch
      templates = []
      for t, sn_best in zip(t_spec_best, sne_best):
          if t == epoch:  templates.append(sn_best)
      templates = list(set(templates))
      
      for template in templates:
          
          phases, err_phases = np.array([]), np.array([])
          for t, sn_best, spec_phase, err_spec_phase in zip(t_spec_best, sne_best, t0s_best, rms_t0s_best):
              if t == epoch and sn_best == template:
                  phases     = np.append(phases, spec_phase)
                  err_phases = np.append(err_phases, err_spec_phase)

          t0_best, rms_t0_best =  weighted_average(phases, err_phases)
          new_sne_best.append(template)
          new_t0_best.append(t0_best)
          new_rms_t0_best.append(rms_t0_best*np.sqrt(float(len(phases))))
  
  sne_best, t0_best, rms_t0_best = np.array(new_sne_best), np.array(new_t0_best), np.array(new_rms_t0_best)
  
  return sne_best, t0_best, rms_t0_best

src/SNII_ETOS.py
import numpy as np
from snid_routines import snid_results
from pdf_routines import typical_pdf_per_sn_bm_per_spectrum, average_pdf_per_sn_bm, average_pdf_per_sn_bm_with_t0_error, final_pdfs, values_from_distribution
from plot_ETOS import plot_ETOS

class Explosion_Time_from_Optical_Spectra:
  def __init__(self, templates_path, sn, z, JD_spectra, wl_spectra, f_spectra, JD_fd, JD_ln=-9999, snid_verbose=False, display_snid_plots=False):
    
    #transform to the rest frame
    JD, wl, f = [], [], []
    for JD_spectrum, wl_spectrum, f_spectrum in zip(JD_spectra, wl_spectra, f_spectra):
        
        wl_spectrum = wl_spectrum/(1.0+z)
        f_spectrum  = f_spectrum*(1.0+z)
        
        #discard spectra at >40 days since the first detection
        if (JD_spectrum - JD_fd) / (1.0+z) >= 40.0:
            print("discarding spectrum at JD="+str(JD_spectrum)+": it is at >40 days since the first detection")
        else:
            JD.append(JD_spectrum)
            wl.append(wl_spectrum)
            f.append(f_spectrum)
            if min(wl_spectrum) > 4100.0:  print("WARNING: spectrum at JD="+str(JD_spectrum)+" has a minimum wavelength higher than the requested (<=410 nm)")
            if max(wl_spectrum) < 7000.0:  print("WARNING: spectrum at JD="+str(JD_spectrum)+" has a maximum wavelength lower than the requested (>=700 nm)")
    JD_spectra, wl_spectra, f_spectra = JD, wl, f
    

    sne_bm, t0s_bm, rms_t0s_bm = {}, {}, {}
    
    #1) run snid for each of the input spectra
    t_spec_bm, sne_bm['1'], t0s_bm['1'], rms_t0s_bm['1'] = snid_results(templates_path, sn, z, JD_fd, JD_spectra, wl_spectra, f_spectra, snid_verbose, display_snid_plots)

    if len(t0s_bm['1']) == 0:
        print("WARNING: Explosion_Time_from_Optical_Spectra without results!")
    else:
        n_spec = len(list(set(t_spec_bm)))
        if n_spec == 1: rms_t0 = 5.0
        if n_spec >  1: rms_t0 = 4.1
        xmin, xmax, dx = -55.0, 55.0, 0.01
        x_pdf = np.linspace(xmin, xmax, int(round((xmax-xmin)/dx,0))+1)
        N_sample = 200000
    
        np.random.seed(1)
        
        #2) select the typical pdf per SN per input spectrum
        sne_bm['2'], t0s_bm['2'], rms_t0s_bm['2'] = typical_pdf_per_sn_bm_per_spectrum(t0s_bm['1'], rms_t0s_bm['1'], sne_bm['1'], t_spec_bm)

        #3) compute the average pdf per SN
        sne_bm['3'], t0s_bm['3'], rms_t0s_bm['3'] = average_pdf_per_sn_bm(t0s_bm['2'], rms_t0s_bm['2'], sne_bm['2'])

        #4) include uncertainty due to the t0
        pdfs_snid_sn_bm = average_pdf_per_sn_bm_with_t0_error(sne_bm['3'], t0s_bm['3'], rms_t0s_bm['3'], x_pdf, N_sample)

        #5) compute the final pdf's
        pdf = {}
        pdf['fd+ln'], pdf["snid"], pdf["snid+fd"], pdf["snid+fd+ln"] = final_pdfs(z, JD_ln, JD_fd, pdfs_snid_sn_bm, x_pdf, N_sample, rms_t0)  

        
        JD_t0, rms_JD_t0 = {"fd+ln":{}}, {"fd+ln":{}}
        t0, rms_t0 = {"snid":{}, "snid+fd":{}, "snid+fd+ln":{}}, {"snid":{}, "snid+fd":{}, "snid+fd+ln":{}}
        for pdf_name in ["fd+ln", "snid", "snid+fd", "snid+fd+ln"]:
            
            if pdf_name == "fd+ln":
            
                for method in ['mean', 'median', 'mode']:
                    if JD_ln != -9999:
                        JD_t0[pdf_name][method] = round((JD_ln + JD_fd)*0.5, 2)
                        rms_JD_t0[pdf_name][method] = round((JD_fd - JD_ln)/np.sqrt(12.0), 2)
                    else:
                        JD_t0[pdf_name][method] = -9999
                        rms_JD_t0[pdf_name][method] = -9999
            else:
            
                t0s = values_from_distribution(x_pdf, pdf[pdf_name], N_sample*10)
                
                JD_t0[pdf_name], rms_JD_t0[pdf_name] = {}, {}
                for method in ['mean', 'median', 'mode']:
                    
                    if method == 'mean':
                        t0[pdf_name][method] = np.mean(t0s)
                    if method == 'median':
                        pc_func = np.array([0.0])
                        for i in range(1, len(x_pdf)):
                            integral = (x_pdf[i]-x_pdf[i-1])*(pdf[pdf_name][i]+pdf[pdf_name][i-1])*0.5
                            pc_func = np.append(pc_func, pc_func[i-1]+integral)
                        pc_func = pc_func / max(pc_func)
                        t0[pdf_name][method] = np.interp(0.5, pc_func, x_pdf)
                    if method == 'mode':
                        i_mode          = pdf[pdf_name].tolist().index(max(pdf[pdf_name]))
                        xs_for_mode     = x_pdf[i_mode-2:i_mode+3]
                        t0s_for_mode    = pdf[pdf_name][i_mode-2:i_mode+3]
                        pars            = np.polyfit(xs_for_mode, t0s_for_mode, 2)[::-1]
                        t0[pdf_name][method] = -pars[1]/(2.0*pars[2])
                        
                        if t0["snid"][method] < 0.0:
                            t0["snid+fd"][method] = t0["snid"][method]
                            if t0["snid"][method] > (JD_ln - JD_fd)/(1.0+z):
                                t0["snid+fd+ln"][method] = t0["snid"][method]
                            else:
                                t0["snid+fd+ln"][method] = (JD_ln - JD_fd)/(1.0+z)
                        else:
                            t0["snid+fd"][method] = 0.0
                            t0["snid+fd+ln"][method] = 0.0
                
                    #rms
                    residuals   = t0s - t0[pdf_name][method]
                    rms_t0[pdf_name][method] = np.sqrt(np.sum(residuals**2)/float(len(residuals)-1))
                    #JD_t0 and rms
                    JD_t0[pdf_name][method]     = round(t0[pdf_name][method]*(1.0+z) + JD_fd, 2)
                    rms_JD_t0[pdf_name][method] = round(rms_t0[pdf_name][method]*(1.0+z), 2)
            
        #input values
        self.sn     = sn
        self.z      = z
        self.JD_fd  = JD_fd
        self.JD_ln  = JD_ln
        self.n_spec = n_spec
        #final pdf's
        self.x_pdf           = x_pdf
        self.pdf             = pdf
        #explosion times and rms
        self.JD_t0 = JD_t0
        self.rms_JD_t0 = rms_JD_t0
        #values to construct figures
        self.for_plots = [sne_bm, t0s_bm, rms_t0s_bm, pdfs_snid_sn_bm]
  
  #plot
  def plot(self, figure_name=''):
      
      plot_ETOS(self, figure_name)
  
  #pick N random values from a specific pdf
  def random_values(self, pdf_name, N):
           
      
      if pdf_name not in ["fd+ln", "snid", "snid+fd", "snid+fd+ln"]:
          print('ERROR in random_values: '+pdf_name+' is not valid! Use "fd+ln", "snid", "snid+fd", of "snid+fd+ln".')
      else:
          JD_pdf = self.x_pdf*(1.0+self.z) + self.JD_fd
          JD_t0s = values_from_distribution(JD_pdf, self.pdf[pdf_name], N)
          
          return JD_t0s
  
  #compute the boundaries enclosing a certain per cent of a pdf
  def pc_boundaries(self, pdf_name, method, percent):
      
      if pdf_name not in ["fd+ln", "snid", "snid+fd", "snid+fd+ln"]:
          print('ERROR in pc_boundaries: '+pdf_name+' is not valid! Use "snid", "snid+fd", of "snid+fd+ln".')
      elif method not in ["mean", "median", "mode"]:
          print('ERROR in pc_boundaries: '+method+' is not valid! Use "mean", "median", of "mode".')
      elif percent < 0.0 or percent > 100.0:
          print('ERROR in pc_boundaries: '+str(percent)+' is not valid! Use a value between 0 and 100.')
      else:
          if pdf_name == "fd+ln":
              if self.JD_ln != -9999:
                  JD_l = 0.5*(self.JD_fd*(1.0-percent/100.0)+self.JD_ln*(1.0+percent/100.0))
                  JD_u = 0.5*(self.JD_fd*(1.0+percent/100.0)+self.JD_ln*(1.0-percent/100.0))
              else:
                  JD_l, JD_u = -9999, -9999
          else:
              JD_t0 = self.JD_t0[pdf_name][method]
              pdf   = self.pdf[pdf_name]
              JD_pdf = self.x_pdf*(1.0+self.z) + self.JD_fd
                       
              pc_func = np.array([0.0])
              for i in range(1, len(JD_pdf)):
                  integral = (JD_pdf[i]-JD_pdf[i-1])*(pdf[i]+pdf[i-1])*0.5
                  pc_func = np.append(pc_func, pc_func[i-1]+integral)
              pc_func = pc_func / max(pc_func)
              
              if JD_t0 == self.JD_ln:
                  JD_l = JD_t0
                  JD_u = np.interp(percent/100.0, pc_func, JD_pdf)
              elif JD_t0 == self.JD_fd:
                  JD_u = JD_t0
                  JD_l = np.interp(1.0-percent/100.0, pc_func, JD_pdf)
              else:
                  alpha_c = round(np.interp(JD_t0, JD_pdf, pc_func),3)
                  alpha_l = alpha_c - 0.5*percent/100.0
                  alpha_u = alpha_c + 0.5*percent/100.0
                  if alpha_l <= 0.0:
                      if "ln" in pdf_name:
                          JD_l = self.JD_ln
                      else:
                          JD_l = min(JD_pdf)
                      JD_u = np.interp(percent/100.0, pc_func, JD_pdf)
                  elif alpha_u >= 1.0:
                      JD_u = self.JD_fd
                      JD_l = np.interp(1.0-percent/100.0, pc_func, JD_pdf)
                  else: 
                      JD_l = np.interp(alpha_l, pc_func, JD_pdf)
                      JD_u = np.interp(alpha_u, pc_func, JD_pdf)
              
              if percent == 100.0 and pdf_name in ["snid+fd", "snid+fd+ln"]:
                  JD_u = self.JD_fd
                  if pdf_name == "snid+fd+ln":
                      JD_l = self.JD_ln
              
          boundaries = np.array([round(JD_l,2), round(JD_u,2)])
          
          return boundaries
